src/plugins/mvcc_delete_plugin.cpp

```diff
@@ -0,0 +1,177 @@
+        const CommitID lowest_end_commit_id =
+            *std::min_element(std::begin(chunk->mvcc_data()->end_cids), std::end(chunk->mvcc_data()->end_cids));
+        const CommitID commit_id_diff = TransactionManager::get().last_commit_id() - lowest_end_commit_id;
+        const CommitID max_commit_id_diff =
+            table->max_chunk_size() * static_cast<CommitID>(_delete_threshold_commit_diff_factor);
+
+        // Evaluate metrics
+        const bool criterion1 = _delete_threshold_rate_invalidated_rows <= rate_of_invalidated_rows;
+        const bool criterion2 = max_commit_id_diff <= commit_id_diff;
+        if (criterion1 || criterion2) _delete_chunk(table_name, chunk_id);
+      }
+    }  // for each chunk
+  }    // for each table
+}
+
+/**
+ * This function processes the physical-delete-queue until its empty.
+ */
+void MvccDeletePlugin::_physical_delete_loop() {
+  std::unique_lock<std::mutex> lock(_mutex_physical_delete_queue);
+
+  while (!_physical_delete_queue.empty()) {
+    ChunkSpecifier chunk_spec = _physical_delete_queue.front();
+    bool success = _delete_chunk_physically(chunk_spec.table_name, chunk_spec.chunk_id);
+
+    if (success) {
+      _physical_delete_queue.pop();
+    } else {
+      return;  // wait for more transactions to finish
+    }
+  }
+}
+
+void MvccDeletePlugin::_delete_chunk(const std::string& table_name, const ChunkID chunk_id) {
+  // Delete chunk logically
+  bool success = _delete_chunk_logically(table_name, chunk_id);
+
+  // Queue physical delete
+  if (success) {
+    DebugAssert(StorageManager::get().get_table(table_name)->get_chunk(chunk_id)->get_cleanup_commit_id() !=
+                    MvccData::MAX_COMMIT_ID,
+                "Chunk needs to be deleted logically before deleting it physically.")
+
+        std::unique_lock<std::mutex>
+            lock(_mutex_physical_delete_queue);
+    _physical_delete_queue.emplace(table_name, chunk_id);
+  } else {
+    std::cout << "Logical delete of chunk " << chunk_id << " failed." << std::endl;
+  }
+}
+
+bool MvccDeletePlugin::_delete_chunk_logically(const std::string& table_name, const ChunkID chunk_id) {
+  const auto& table = StorageManager::get().get_table(table_name);
+  const auto& chunk = table->get_chunk(chunk_id);
+
+  DebugAssert(chunk != nullptr, "Chunk does not exist. Physical Delete can not be applied.")
+      DebugAssert(chunk_id < (table->chunk_count() - 1),
+                  "MVCC Logical Delete should not be applied on the last/current mutable chunk.")
+
+      // Create temporary referencing table that contains the given chunk only
+      auto transaction_context = TransactionManager::get().new_transaction_context();
+  auto gt = std::make_shared<GetTable>(table_name);
+  gt->set_transaction_context(transaction_context);
+
+  auto chunk_count = table->chunk_count();
+  std::vector<ChunkID> excluded_chunk_ids;
+  for (ChunkID excluded_chunk_id = ChunkID{0}; excluded_chunk_id < chunk_count; ++excluded_chunk_id) {
```

Comment:
I'm wondering if it would be easier to take all chunk_ids and then remove `chunk_id` from this vector

Reply:
Hmm, I get your point, but then we still would have to fill a vector with the chunk ids first, and remove one of the values afterwards. What do you think about an approach like this (iota just fills up a vector with ascending numbers)?
```c++
std::vector<ChunkID> excluded_chunk_ids(table->chunk_count());
std::iota(std::begin(excluded_chunk_ids), std::end(excluded_chunk_ids), 0);
excluded_chunk_ids.erase(std::remove(std::begin(excluded_chunk_ids), std::end(excluded_chunk_ids), chunk_id));
```